### PR TITLE
CORE-4211 Fixes to help with error handling

### PR DESCRIPTION
**Pull request by michael-redpanda**

### CODE REVIEW

**Comment by michael-redpanda** [comment]
Changes being tested here: https://github.com/redpanda-data/vtools/pull/2858

File: src/net/ossl.cc [hunk 1]
```diff
@@ -814,17 +815,17 @@ class session : public enable_shared_from_this<session>, public session_impl {
                 _eof = true;
                 return make_ready_future<stop_iteration>(stop_iteration::yes);
             }
-            auto err = make_exception_ptr(ossl_error::make_ossl_error(
-                "Error occurred during SSL write"));
+            auto err = ossl_error::make_ossl_error(
+                "Error occurred during SSL write");
             return handle_output_error(err).then([] {
```
Comment (oleiman) [inline src/net/ossl.cc:828]: question: should these be `std::move(err)` something?

File: src/net/ossl.cc [hunk 2]
```diff
@@ -802,8 +802,16 @@ class session : public enable_shared_from_this<session>, public session_impl {
...
+            auto f = make_ready_future<>();
+            if (errno == 0) {
+                auto err = ossl_error::make_ossl_error("System error encountered during SSL write");
+                f = handle_output_error(err);
+            } else {
+                auto err = std::system_error(
+                    errno, std::system_category(), "System error encountered during SSL write");
+                f = handle_output_error(err);
+            }
+            return f.then([] {
```
Comment (oleiman) [inline src/net/ossl.cc:814]: suggestion (non-blocking): not sure it's any cleaner, really, but maybe...

```suggestion
            [errno]() -> future<> {
                if (errno == 0) {
                    auto err = ossl_error::make_ossl_error("System error encountered during SSL write");
                    return handle_output_error(err);
                } else {
                    auto err = std::system_error(
                        errno, std::system_category(), "System error encountered during SSL write");
                    return handle_output_error(err);
                }
            }().then([] {
```

File: src/net/ossl.cc [hunk 3]
```diff
@@ -1068,10 +1081,12 @@ class session : public enable_shared_from_this<session>, public session_impl {
                 case SSL_ERROR_SYSCALL:
                     // check to see if errno is 0, which may indicate some sort of shutdown
                     if (errno == 0) {
-                        return make_ready_future<buf_type>();
+                        _error = std::make_exception_ptr(ossl_error::make_ossl_error(
+                            "System error during SSL read"));
+                    } else {
+                        _error = std::make_exception_ptr(std::system_error(
+                        errno, std::system_category(), "System error during SSL read"));
```
Comment (oleiman) [inline src/net/ossl.cc:1088]: nit: questionable formatting (left justification of the second line...)

**Review by oleiman — COMMENTED** [review]
makes sense; looks good. few little things, possibly all no-op.

File: src/net/ossl.cc (hunk 1 above)
Comment (michael-redpanda) [reply src/net/ossl.cc:828]: Probably not a bad idea to save some space

**Review by oleiman — APPROVED** [review]
lgtm

File: src/net/ossl.cc [hunk 4]
```diff
@@ -1824,8 +1845,9 @@ const int seastar::tls::ERROR_UNKNOWN_SRP_USERNAME = ERR_PACK(
   ERR_LIB_SSL, 0, SSL_R_INVALID_SRP_USERNAME);
 const int seastar::tls::ERROR_PREMATURE_TERMINATION = ERR_PACK(
   ERR_LIB_SSL, 0, SSL_R_UNEXPECTED_EOF_WHILE_READING);
-const int seastar::tls::ERROR_PUSH = ERR_PACK(
-  ERR_LIB_SSL, 0, SSL_R_BIO_NOT_SET);
+// System errors are not ERR_PACK'ed like other errors but instead
+// are OR'ed with ((unsigned int)INT_MAX + 1)
+const int seastar::tls::ERROR_PUSH = int(ERR_SYSTEM_FLAG | EPIPE);
```
Comment (pgellert) [inline src/net/ossl.cc:1850]: question: can we ever see the old error `ERR_PACK(ERR_LIB_SSL, 0, SSL_R_BIO_NOT_SET)` as a push error or is that not possible anymore?

Also, just to double check, this is intentionally `int(ERR_SYSTEM_FLAG | EPIPE)` and not `ERR_PACK(ERR_LIB_???, 0, ERR_SYSTEM_FLAG | EPIPE)`, right?

**Review by pgellert — APPROVED** [review]

File: src/net/ossl.cc (hunk 4 above)
Comment (michael-redpanda) [reply src/net/ossl.cc:1850]: > question: can we ever see the old error ERR_PACK(ERR_LIB_SSL, 0, SSL_R_BIO_NOT_SET) as a push error or is that not possible anymore?

Fair question.  The push/pull errors are more artifacts of GnuTLS.  This was a best approximation I could do with what I think the `PULL` error on GnuTLS means with OpenSSL

> Also, just to double check, this is intentionally int(ERR_SYSTEM_FLAG | EPIPE) and not ERR_PACK(ERR_LIB_???, 0, ERR_SYSTEM_FLAG | EPIPE), right?

Yes.  When an `errno` is raised within OpenSSL, it is paired with `ERR_LIB_SYS` however OpenSSL drops the `lib` portion and just or's the errno with `ERR_SYSTEM_FLAG`

Comment (pgellert) [reply src/net/ossl.cc:1850]: Thanks, makes sense 👍

**Comment by michael-redpanda** [comment]
There was one last unhandled error in the final CDT run.  Created an issue for follow up: https://redpandadata.atlassian.net/browse/CORE-4225

**Review by BenPope — APPROVED** [review]
LGTM, but I think it can be simplified overall: https://github.com/michael-redpanda/seastar/compare/CORE-4211-Fix-nested-exception...BenPope:seastar:CORE-4211-Fix-nested-exception?expand=1

**Comment by michael-redpanda** [comment]
Force push `9910e86`:

* Took @BenPope 's suggestions
* Updated dependencies

**Comment by michael-redpanda** [comment]
Force push `ebb5558`:

* Added missing newline at end of `tls.hh`

File: src/net/ossl.cc [hunk 5]
```diff
@@ -776,7 +784,8 @@ class session : public enable_shared_from_this<session>, public session_impl {
             return make_ready_future<stop_iteration>(stop_iteration::no);
         case SSL_ERROR_SYSCALL:
         {
-            auto err = std::system_error(errno, std::system_category(), "System error encountered during SSL write");
+            auto err = make_error_from_errno("System error encountered during SSL write");
+            auto f = make_ready_future<>();
```
Comment (BenPope) [inline src/net/ossl.cc:788]: `f` is unused

File: install-dependencies.sh [hunk 6]
```diff
@@ -224,6 +226,7 @@ opensuse_packages=(
     libtool
     stow
     openssl
+    openssl-devel
```
Comment (BenPope) [inline install-dependencies.sh:229]:
```suggestion
    libssl-dev

**Comment by michael-redpanda** [comment]
Force push `fa1ef72`:

* Fixups

**Review by BenPope — APPROVED** [review]

**Review by BenPope — APPROVED** [review]

**Comment by michael-redpanda** [comment]
Force push `f9b9a82`:

Dropped change in dependencies as that was causing issues with RP build - will address in separate PR

**Review by BenPope — APPROVED** [review]

**Review by oleiman — APPROVED** [review]

**Pull request merged**